kubectl-volsync/cmd/migration_rsync.go

```diff
@@ -17,36 +17,207 @@ along with this program. If not, see <http://www.gnu.org/licenses/>.
+	client client.Client
+}
+
+// migrationCreateCmd represents the create command
+var migrationSyncCmd = &cobra.Command{
 	Use:   "rsync",
-	Short: "A brief description of your command",
-	Long: `A longer description that spans multiple lines and likely contains examples
-and usage of using your command. For example:
-
-Cobra is a CLI library for Go that empowers applications.
-This application is a tool to generate the needed files
-to quickly create a Cobra application.`,
-	Run: func(cmd *cobra.Command, args []string) {
-		fmt.Println("rsync called")
+	Short: i18n.T("Rsync data from source to destination"),
+	Long: templates.LongDesc(i18n.T(`
+	This command ensures the migration of data from source to destination
+	via rsync over ssh. The execution of this command should be followed by
+	migration create which establishes the relationship.
+	`)),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		ms := &migrationSync{}
+		mr, err := loadMigrationRelationship(cmd)
+		if err != nil {
+			return err
+		}
+		ms.mr = mr
+
+		err = ms.newMigrationSync(cmd)
+		if err != nil {
+			return err
+		}
+
+		return ms.Run(cmd.Context())
 	},
 }
 
 func init() {
-	migrationCmd.AddCommand(migrationRsyncCmd)
+	initmigrationSyncCmd(migrationSyncCmd)
+}
+
+func initmigrationSyncCmd(migrationCreateCmd *cobra.Command) {
+	migrationCmd.AddCommand(migrationCreateCmd)
+
+	migrationCreateCmd.Flags().String("source", "", "source volume to be migrated")
+	cobra.CheckErr(migrationCreateCmd.MarkFlagRequired("source"))
+}
+
+func (ms *migrationSync) Run(ctx context.Context) error {
+	mrd := ms.mr.data.Destination
+	k8sClient, err := newClient(mrd.Cluster)
+	if err != nil {
+		return err
+	}
+	ms.client = k8sClient
+
+	// Ensure source volume
+	_, err = os.Stat(ms.Source)
+	if err != nil {
+		return fmt.Errorf("failed to access the source volume, %w", err)
+	}
+
+	var sshKeyDir *string
+	defer func() {
+		// Remove the directory containing secrets
+		if sshKeyDir != nil {
+			if err = os.RemoveAll(*sshKeyDir); err != nil {
+				klog.Infof("failed to remove temporary directory with ssh keys (%s): %w",
+					sshKeyDir, err)
+			}
+		}
+	}()
+	// Retrieve Secrets/keys
+	sshKeyDir, err = ms.retrieveSecrets(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Do rysnc
+	err = ms.runRsync(ctx, *sshKeyDir)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func loadMigrationRelationship(cmd *cobra.Command) (*migrationRelationship, error) {
+	r, err := LoadRelationshipFromCommand(cmd, MigrationRelationshipType)
+	if err != nil {
+		return nil, err
+	}
+
+	mr := &migrationRelationship{
+		Relationship: *r,
+	}
+
+	// Decode according to the file version
+	version := mr.GetInt("data.version")
+	switch version {
+	case 1:
+		if err := mr.GetData(&mr.data); err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("unsupported config file version %d", version)
+	}
+	return mr, nil
+}
+
+func (ms *migrationSync) newMigrationSync(cmd *cobra.Command) error {
```

Comment:
I think something like below would be better, So that we can just pass `ctx` to `Run` like `migration_create`

```
RunE: func(cmd *cobra.Command, args []string) error {
		ms, err := newMigrationSync(cmd)
		if err != nil {
			return err
		}
		mr, err := loadMigrationRelationship(cmd)
		if err != nil {
			return err
		}
		ms.mr = mr

		return ms.Run(cmd.Context())
	},
```

Reply:
Yes. The goal is to have the same structure for all the commands.

Reply:
sure